pallets/account/src/impls.rs

```diff
@@ -0,0 +1,174 @@
+    fn on_killed_account(who: &T::AccountId) {
+        // remove mapping created by `claim_account` or `get_or_create_evm_address`
+        if let Some(evm_addr) = EvmAccounts::<T>::get(who) {
+            NativeAccounts::<T>::remove(evm_addr);
+            EvmAccounts::<T>::remove(who);
+        }
+    }
+}
+
+/// A lookup implementation returning the `AccountId` from `MultiAddress::Address20` (EVM Address).
+impl<T: Config> StaticLookup for Pallet<T> {
+    type Source = MultiAddress<T::AccountId, ()>;
+    type Target = T::AccountId;
+
+    fn lookup(a: Self::Source) -> Result<Self::Target, LookupError> {
+        match a {
+            MultiAddress::Address20(i) => Ok(
+                <Self as AddressManager<T::AccountId, EvmAddress>>::to_account_id_or_default(
+                    &EvmAddress::from_slice(&i),
+                ),
+            ),
+            _ => Err(LookupError),
+        }
+    }
+
+    fn unlookup(a: Self::Target) -> Self::Source {
+        MultiAddress::Id(a)
+    }
+}
+
+/// EIP-712 compatible signature scheme for verifying ownership of EVM Address
+///
+/// Raw Data = Domain Separator + Type Hash + keccak256(AccountId)
+pub struct EIP712Signature<T: Config, ChainId: Get<u64>>(PhantomData<(T, ChainId)>);
+impl<T: Config, ChainId: Get<u64>> ClaimSignature for EIP712Signature<T, ChainId> {
+    type AccountId = T::AccountId;
+    /// EVM address type
+    type Address = EvmAddress;
+    /// A signature (a 512-bit value, plus 8 bits for recovery ID).
+    type Signature = [u8; 65];
+
+    fn build_signing_payload(who: &Self::AccountId) -> [u8; 32] {
+        let domain_separator = Self::build_domain_separator();
+        let args_hash = Self::build_args_hash(who);
+
+        let mut payload = b"\x19\x01".to_vec();
+        payload.extend_from_slice(&domain_separator);
+        payload.extend_from_slice(&args_hash);
+        keccak_256(&payload)
+    }
+
+    fn verify_signature(who: &Self::AccountId, sig: &Self::Signature) -> Option<EvmAddress> {
+        let payload_hash = Self::build_signing_payload(who);
+
+        sp_io::crypto::secp256k1_ecdsa_recover(sig, &payload_hash)
+            .map(|pubkey| H160::from(H256::from_slice(&keccak_256(&pubkey))))
+            .ok()
+    }
+}
+
+impl<T: Config, ChainId: Get<u64>> EIP712Signature<T, ChainId> {
+    /// TODO: minor, use hardcoded bytes, configurable via generics
+    fn build_domain_separator() -> [u8; 32] {
+        let mut hash =
+            keccak256!("EIP712Domain(string name,string version,uint256 chainId,bytes32 salt)")
+                .to_vec();
+        hash.extend_from_slice(&keccak256!("Astar EVM Claim")); // name
+        hash.extend_from_slice(&keccak256!("1")); // version
+        hash.extend_from_slice(&(<[u8; 32]>::from(U256::from(ChainId::get())))); // chain id
+        hash.extend_from_slice(
+            frame_system::Pallet::<T>::block_hash(T::BlockNumber::zero()).as_ref(),
+        ); // genesis block hash
+        keccak_256(hash.as_slice())
```

Comment:
It's unclear to what _time_ does the TODO refer to?

If it's not going to be implemented now, please document the data scheme above the `EIP721Signature` struct.

For the strings being hashed, I'd suggest to avoid spaces. Use underscore instead, and/or remove the variable names from the function signatures (more EVM-like).

Also, about the usage of `#[eip712(...` in tests - cannot it be reused here too?

Reply:
It's a very minor optimisation which can worked upon later.

>if it's not going to be implemented now, please document the data scheme above the EIP721Signature struct.

The data scheme for EIP712 Domain Hash is very well documented, I'll add the link to the EIP doc in comments.

> Also, about the usage of #[eip712(... in tests - cannot it be reused here too?

It can but I'd like not to because,
- since macro only take hex literal for salt and we are using genesis hash and it's very difficult to hardcoded it for every chain env (zombienet, tests, benchmarks)
- It'll add a lot of unnecessary dependencies (`ethers.rs` is very big package) into our runtime which can be avoided by simply building it manually.

Reply:
> The data scheme for EIP712 Domain Hash is very well documented, I'll add the link to the EIP doc in comments.
Right, I went over the specification after posting the initial review.
Linking it sounds good. 👍 

> It's a very minor optimisation which can worked upon later.
I don't see it as optimization but as security.
Code reuse is better than implementing something by ourselves.

> It'll add a lot of unnecessary dependencies (ethers.rs is very big package) into our runtime which can be avoided by simply building it manually.

Perhaps - it's worth it though IMHO.
If something exists that's well implemented & tested, no need to reinvent the wheel.
Especially with something as "sensitive" as this.

If it brings a significant size increase to the runtime blob (optimized build, compact & compressed binary), then we should at least make sure to have more tests that compare data generated by `ethers` lib and our custom code.

Reply:
I don't mean to reinvent the wheel but it's not a very complex task to build the signing payload manually, it's only 6-10 lines for domain hash and including a library to do that same with more than 1K+ lines doesn't feel justified.

We do have a test for checking the payload generation against ethers.rs and I don't think we need more tests to check the domain and type hash since they are constant values (only depending on genesis hash). And overall payload generation is just concat of keccack hash of types and value, that's it.
We can have more tests if that is required, but I'm not sure what do we want to check more?

Reply:
I'm aware what it is, and that it's just a few lines of code.
I still don't think you should "just" rewrite that code on your own.

Doesn't matter how much some lib has LoCs, if it already solves a problem you need solved.
My guess is that lib is much better tested than your few lines of code.

Anyhow, feel free to get others opinions too.

Reply:
@ashutoshvarma Could you try to build locally and see how much extra size it will add to the runtime blob?

Reply:
@Dinonard @shaunxw 
`ethers.rs` does not support `no_std` - https://github.com/gakonst/ethers-rs/issues/2560.
Another big library `rust-web3` also does not support `no_std` - https://github.com/tomusdrw/rust-web3/issues/21 (issue by Rob)

There are several other small libraries for eip712 but not sure if anyone is using them and they are well tested
- https://github.com/paritytech/EIP-712 - archived 4 years ago
- https://github.com/graphprotocol/eip-712-derive - no readme, no docs